gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/RefreshTokenAuth2Provider.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2022 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.hadoop.fs.gcs;
+
+import static com.google.api.client.util.Preconditions.checkNotNull;
+import static com.google.common.flogger.LazyArgs.lazy;
+
+import com.google.api.client.auth.oauth2.ClientParametersAuthentication;
+import com.google.api.client.auth.oauth2.RefreshTokenRequest;
+import com.google.api.client.auth.oauth2.TokenRequest;
+import com.google.api.client.auth.oauth2.TokenResponse;
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.cloud.hadoop.gcsio.GoogleCloudStorageOptions;
+import com.google.cloud.hadoop.util.AccessTokenProvider;
+import com.google.cloud.hadoop.util.HttpTransportFactory;
+import com.google.cloud.hadoop.util.RedactedString;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.flogger.GoogleLogger;
+import java.io.IOException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Optional;
+import org.apache.hadoop.conf.Configuration;
+
+/**
+ * Retrieve an access token using the OAuth2 refresh token grant flow. See <a
+ * href="https://datatracker.ietf.org/doc/html/rfc6749#section-1.5">RFC 6749</a>.
+ */
+public class RefreshTokenAuth2Provider implements AccessTokenProvider {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+  private static final DateTimeFormatter dateFormat =
+      DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS").withZone(ZoneId.systemDefault());
```

Comment:
Is date format so important for logging, that we can not rely on default `toString` logic of Java date classes?

Reply:
I am going to double check but  it's kind of is: Troubleshooting the expiration time of the token can be challenging, especially if like us, we don't have access to our customers hadoop cluster.

Reply:
I checked: Yeah it's fine, the `toString` of instant is readable enough:

```
dateFormat.format(Instant.ofEpochMilli(accessToken.get().getExpirationTimeMilliSeconds())) + " VS " + Instant.ofEpochMilli(accessToken.get().getExpirationTimeMilliSeconds())
```

=> 

```
2022-01-18 12:01:32.539 VS 2022-01-18T11:01:32.539Z
```